### PR TITLE
Upload scans

### PR DESCRIPTION
**Pull request by jssblck**
# Overview

Uploads completed scans to FOSSA.

## Acceptance criteria

When scans are complete, they are able to be uploaded.

## Testing plan

Unfortunately I don't have a good automated test solution for this, since this is all effectful.
I'd like to add this in a future PR, but for now I relied on manual testing.

I configured the following:
```
fossa_endpoint: https://app.fossa.com
fossa_integration_key: <key>
version: 1

debugging:
  location: /Users/jessica/.config/fossa/broker/debug

integrations:
  - type: git
    poll_interval: 1h
    remote: git@github.com:fossas/broker.git
    auth:
      type: ssh_key_file
      path: ~/.ssh/id_rsa
```

I then ran Broker and got the following output (trimmed to just the INFO lines):
```
Finished dev [unoptimized + debuginfo] target(s) in 7.44s
2023-04-04T00:31:54.704424Z  INFO broker::debug: Debug artifacts being stored in '/Users/jessica/.config/fossa/broker/debug/trace'
2023-04-04T00:31:54.731451Z  INFO broker::subcommand::run: Polling 'git:ssh:git@github.com:fossas/broker.git'
2023-04-04T00:31:59.325167Z  INFO broker::subcommand::run: Enqueued task to scan 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:broker-fix@61fd8c768c6bdf42a2c172e802c1d1dacda830a5'
2023-04-04T00:31:59.325323Z  INFO broker::subcommand::run: Enqueued task to scan 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:main@fd5dda71a2e9a6da827a9319257dc59ee009128f'
2023-04-04T00:31:59.325462Z  INFO broker::subcommand::run: Enqueued task to scan 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:test-with-async_with_vars@98f35745378eaaf375b0df6e055e7166077c677f'
2023-04-04T00:31:59.325590Z  INFO broker::subcommand::run: Enqueued task to scan 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:upload-scans@3cf743d16a8fa84007fa0515aad8c5ffaa07014b'
2023-04-04T00:31:59.325683Z  INFO broker::subcommand::run: Next poll interval for 'git:ssh:git@github.com:fossas/broker.git' in 3600s
2023-04-04T00:31:59.326970Z  INFO broker::subcommand::run: Scanning 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:broker-fix@61fd8c768c6bdf42a2c172e802c1d1dacda830a5'
2023-04-04T00:32:04.947601Z  INFO broker::subcommand::run: Scanned 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:broker-fix@61fd8c768c6bdf42a2c172e802c1d1dacda830a5'
2023-04-04T00:32:04.955194Z  INFO broker::subcommand::run: Scanning 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:main@fd5dda71a2e9a6da827a9319257dc59ee009128f'
2023-04-04T00:32:04.959560Z  INFO broker::subcommand::run: Uploading project: 'git@github.com:fossas/broker.git@61fd8c768c6bdf42a2c172e802c1d1dacda830a5 (broker-fix)'
2023-04-04T00:32:05.359134Z  INFO broker::subcommand::run: Uploaded project 'git@github.com:fossas/broker.git@61fd8c768c6bdf42a2c172e802c1d1dacda830a5 (broker-fix)' as locator: 'custom+24357/git@github.com:fossas/broker.git$61fd8c768c6bdf42a2c172e802c1d1dacda830a5'
2023-04-04T00:32:08.689091Z  INFO broker::subcommand::run: Scanned 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:main@fd5dda71a2e9a6da827a9319257dc59ee009128f'
2023-04-04T00:32:08.696913Z  INFO broker::subcommand::run: Scanning 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:test-with-async_with_vars@98f35745378eaaf375b0df6e055e7166077c677f'
2023-04-04T00:32:08.701382Z  INFO broker::subcommand::run: Uploading project: 'git@github.com:fossas/broker.git@fd5dda71a2e9a6da827a9319257dc59ee009128f (main)'
2023-04-04T00:32:09.079042Z  INFO broker::subcommand::run: Uploaded project 'git@github.com:fossas/broker.git@fd5dda71a2e9a6da827a9319257dc59ee009128f (main)' as locator: 'custom+24357/git@github.com:fossas/broker.git$fd5dda71a2e9a6da827a9319257dc59ee009128f'
2023-04-04T00:32:12.660118Z  INFO broker::subcommand::run: Scanned 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:test-with-async_with_vars@98f35745378eaaf375b0df6e055e7166077c677f'
2023-04-04T00:32:12.667690Z  INFO broker::subcommand::run: Scanning 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:upload-scans@3cf743d16a8fa84007fa0515aad8c5ffaa07014b'
2023-04-04T00:32:12.672192Z  INFO broker::subcommand::run: Uploading project: 'git@github.com:fossas/broker.git@98f35745378eaaf375b0df6e055e7166077c677f (test-with-async_with_vars)'
2023-04-04T00:32:13.011553Z  INFO broker::subcommand::run: Uploaded project 'git@github.com:fossas/broker.git@98f35745378eaaf375b0df6e055e7166077c677f (test-with-async_with_vars)' as locator: 'custom+24357/git@github.com:fossas/broker.git$98f35745378eaaf375b0df6e055e7166077c677f'
2023-04-04T00:32:16.252687Z  INFO broker::subcommand::run: Scanned 'git:ssh:git@github.com:fossas/broker.git' at 'git:branch:upload-scans@3cf743d16a8fa84007fa0515aad8c5ffaa07014b'
2023-04-04T00:32:16.261477Z  INFO broker::subcommand::run: Uploading project: 'git@github.com:fossas/broker.git@3cf743d16a8fa84007fa0515aad8c5ffaa07014b (upload-scans)'
2023-04-04T00:32:16.592327Z  INFO broker::subcommand::run: Uploaded project 'git@github.com:fossas/broker.git@3cf743d16a8fa84007fa0515aad8c5ffaa07014b (upload-scans)' as locator: 'custom+24357/git@github.com:fossas/broker.git$3cf743d16a8fa84007fa0515aad8c5ffaa07014b'
```

I then spot checked the project [here](https://app.fossa.com/projects/custom%2B24357%2Fgit%40github.com%3Afossas%2Fbroker.git/refs/branch/main/fd5dda71a2e9a6da827a9319257dc59ee009128f) (in the `jessoteric` org) and it looked valid.

## References

https://fossa.atlassian.net/browse/ANE-833

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).

### CODE REVIEW

File: src/api/fossa.rs [hunk 1]
```diff
@@ -84,3 +210,216 @@ impl TryFrom<String> for Key {
...
+        .await
+        .change_context_lazy(|| Error::upload_scan(&locator, &source_units))?
+        .into()
+}
+
+impl Endpoint {
+    /// Make a GET request against the FOSSA server with the provided route,
+    /// which is joined to the base.
+    #[tracing::instrument]
+    async fn get<T: DeserializeOwned>(&self, route: &str, token: &Key) -> Result<T, Error> {
```
Comment (spatten) [inline src/api/fossa.rs:348]: I think should to refactor `broker fix` to use this.

Comment (spatten) [reply src/api/fossa.rs:348]: Although, if `lookup` is never used, is this ever used?

**Review by spatten — APPROVED** [review]

File: Makefile [hunk 2]
```diff
@@ -27,4 +27,7 @@ migrate-up:
 migrate-down:
 	@cargo sqlx migrate revert --source db/migrations
 
-.PHONY: test run build dev review-snapshots generate-dist migration-status migrate-up migrate-down
+doc:
+	@cargo doc --open --no-deps
```
[the remark this answers is not in the document]
Comment (jssblck) [reply Makefile:31]: Done! https://fossa.atlassian.net/browse/ANE-910

File: src/api/fossa.rs [hunk 3]
```diff
@@ -84,3 +210,216 @@ impl TryFrom<String> for Key {
...
+    key: Key,
+
+    /// The ID of the organization to which the API key is registered.
+    organization_id: usize,
+}
+
+impl OrgConfig {
+    /// Lookup the organization for the provided config.
+    #[tracing::instrument]
+    pub async fn lookup(config: &Config) -> Result<Self, Error> {
```
[the remark this answers is not in the document]
Comment (jssblck) [reply src/api/fossa.rs:231]: It's not, it was earlier until I realized we don't need to do this. But the code is written, so...

File: src/api/fossa.rs (hunk 1 above)
Comment (jssblck) [reply src/api/fossa.rs:348]: It's not, it was earlier until I realized we don't need to do this. But the code is written, so...

**Pull request merged**